staging/src/k8s.io/api/networking/v1/generated.proto

```diff
@@ -415,7 +415,9 @@ message NetworkPolicyPort {
   optional int32 endPort = 3;
 }
 
-// NetworkPolicySpec provides the specification of a NetworkPolicy
+// NetworkPolicySpec provides the specification of a NetworkPolicy.
+// In order for a connection to succeed, it must be allowed by any egress rules applied to
+// the source pods as well as any ingress rules applied to the destination pods.
```

Comment:
This is ambiguous. If we're concerned about people not understanding the existing documentation, then we have to be concerned about people not understanding this, and assuming that "must be allowed by any egress rules" means that every rule must allow the connection, not that at least one rule must allow the connection. (ie, "∀ rule, rule allows the connection", rather than "∃ rule s.t. rule allows the connection")

You'd need to say: "In order for a connection to succeed, the source pod must either be unisolated with respect to egress, or else must be selected by at least one NetworkPolicy containing an egress rule allowing the connection; and the destination pod must either be unisolated with respect to ingress, or else must be selected by at least one NetworkPolicy containing an ingress rule allowing the connection."

There is no way to explain NP's semantics here that is both concise and non-misleading, because NP's semantics are terrible to explain.

Reply:
> There is no way to explain NP's semantics here that is both concise and non-misleading, because NP's semantics are terrible to explain.

Agree 😄 

Returning to the original issue that prompted this PR, the goal is to highlight some behavior which to both me and @NillsF is surprising and inconsistent with the rest of network policies -- and to at least point the user in the right direction (if it's not possible to give a full, complete, and correct explanation 😄 ).

So in my mind the question is -- is there any way that we can add something brief, helpful, and not wrong that increases the chances that 

 1. users even understand that this is a thing, and so aren't caught unawares
 2. users have some idea of how it works (and can look up additional details elsewhere if necessary -- for example, in @NillsF 's docs PR 💯 )
 
 And if the answer to that question is no, then the follow-up question is: should we just close this PR without merging ? 😄

Reply:
So, you could say something "it must not be blocked by any egress policy or any ingress policy" except that then you're encouraging people to think of NetworkPolicies as blocking traffic rather than allowing traffic, so it fails at "non-misleading".

Having more detailed documentation for network plugin developers would definitely be great, but like Casey said before, I'm not sure the API docs are the right place for that.

At any rate, your newly-added e2e test should prevent anyone else from making this same mistake in the future, so maybe we don't need to say anything in the docs here?

Reply:
Sounds like you're leaning towards closing out this PR @danwinship ?  Fine by me 👍